internal/ociwrapper/ociwrapper.go

```diff
@@ -0,0 +1,254 @@
+package ociwrapper
+
+import (
+	"bytes"
+	"context"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path/filepath"
+	"strconv"
+
+	"github.com/opencontainers/go-digest"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/rancher/fleet/internal/manifest"
+	"oras.land/oras-go/v2"
+	"oras.land/oras-go/v2/content"
+	"oras.land/oras-go/v2/content/memory"
+	"oras.land/oras-go/v2/registry/remote"
+	"oras.land/oras-go/v2/registry/remote/auth"
+	"oras.land/oras-go/v2/registry/remote/retry"
+)
+
+const (
+	fileType     = "application/fleet.file"
+	artifactType = "application/fleet.manifest"
+
+	OCISecretUsername  = "username"
+	OCISecretPassword  = "password"
+	OCISecretURL       = "url"
+	OCISecretBasicHTTP = "http"
+	OCISecretInsecure  = "insecure"
+)
+
+type OCIOpts struct {
+	URL             string
+	Username        string
+	Password        string
+	BasicHTTP       bool
+	InsecureSkipTLS bool
+}
+
+type OrasOpsWrapper interface {
+	PackManifest(ctx context.Context, pusher content.Pusher, packManifestVersion oras.PackManifestVersion, artifactType string, opts oras.PackManifestOptions) (ocispec.Descriptor, error)
+	Copy(ctx context.Context, src oras.ReadOnlyTarget, srcRef string, dst oras.Target, dstRef string, opts oras.CopyOptions) (ocispec.Descriptor, error)
+	NewStore() oras.Target
+}
+
+type OrasOperator struct{}
+
+func (o *OrasOperator) NewStore() oras.Target {
+	return memory.New()
+}
+
+func (o *OrasOperator) PackManifest(ctx context.Context, pusher content.Pusher, packManifestVersion oras.PackManifestVersion, artifactType string, opts oras.PackManifestOptions) (ocispec.Descriptor, error) {
+	return oras.PackManifest(ctx, pusher, packManifestVersion, artifactType, opts)
+}
+
+func (o *OrasOperator) Copy(ctx context.Context, src oras.ReadOnlyTarget, srcRef string, dst oras.Target, dstRef string, opts oras.CopyOptions) (ocispec.Descriptor, error) {
+	return oras.Copy(ctx, src, srcRef, dst, dstRef, opts)
+}
+
+type OCIWrapper struct {
+	oci OrasOpsWrapper
+}
+
+func NewOCIWrapper() *OCIWrapper {
+	return &OCIWrapper{
+		oci: &OrasOperator{},
+	}
+}
+
+func getHTTPClient(insecureSkipTLS bool) *http.Client {
+	if !insecureSkipTLS {
+		return retry.DefaultClient
+	}
+	return &http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: &tls.Config{InsecureSkipVerify: true}, // #nosec G402
+		},
+	}
+}
+
+func getAuthClient(opts OCIOpts) *auth.Client {
+	client := &auth.Client{
+		Client: getHTTPClient(opts.InsecureSkipTLS),
+		Cache:  auth.NewCache(),
+	}
+	if opts.Username != "" {
+		cred := auth.Credential{
+			Username: opts.Username,
+			Password: opts.Password,
+		}
+		client.Credential = func(ctx context.Context, s string) (auth.Credential, error) {
+			return cred, nil
+		}
+	}
+	return client
+}
+
+func newOCIRepository(id string, opts OCIOpts) (*remote.Repository, error) {
+	repo, err := remote.NewRepository(filepath.Join(opts.URL, id))
```

Comment:
I find this problematic. We call it a "URL", but it's a simplified URL at best, as it is missing the protocol part, e.g. "oci://".
Later on we have to modify this string to turn it into  a URL.
ORAS calls this a "reference": https://oras.land/docs/concepts/reference/